### PR TITLE
Update Condition Filter to use OR instead of AND

### PR DESCRIPTION
**Pull request by TurkeyFried**
This PR updates the Condition filters to pass items if any of the filters matches the report's data.

Note - This PR has attached a CURL call that I used on Postman to test. Feel free to use it or make your own test.

Test Steps:
1. In organizations.yml, update the `Ignore FULL_ELR` receiver to have the following condition filter, replacing 600-7 with whatever LOINC code shows up in your Postman test:
`conditionFilter: ["%obsPerformedCodes.intersect('600-7').exists()"]`
2. Run ReportStream, then run`./gradlew reloadTables` and `./gradlew reloadSettings` to update the settings.
3. Submit a report with data that would go to the `Ignore FULL_ELR` Receiver.
4. Using the History API, verify that the report's items were successfully delivered to `Ignore FULL_ELR`.
5. Repeat steps 1-4, but change the `conditionFilter`'s LOINC code to something that is not in the submitted report (e.g.`999-9`). The item should now be filtered out.
6. Add multiple values to `conditionFilter`, with at least one that matches the submission data, and at least one that does not matches. for example,
```
conditionFilter:
        - "%obsPerformedCodes.intersect('999-9').exists()"
        - "%obsPerformedCodes.intersect('600-7').exists()"
        - "%obsPerformedCodes.intersect('123-4').exists()"
```

## Changes
- Update logic on the FHIRRouter to check condition filters in an OR manner instead of AND.

## Linked Issues
- #8837 

[local call curl.txt](https://github.com/CDCgov/prime-reportstream/files/11466673/local.call.curl.txt)

### CODE REVIEW

**Comment by jimmyfagan** [comment]
Oops accidentally got my tabs mixed up and updated this branch with the latest from master, should hopefully be okay, just pointing that out in case.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (jimmyfagan) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:503]: I'd prefer to check for exceptions first, just so we don't keep filters that cause exceptions without realizing it. It would result in the condition filter failing if we find an exception even if it does have a passing condition, but I think that'd be for the best.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (jimmyfagan) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:477]: It seems like we're using this list as more of a boolean, it'll either be empty or it won't, but we're never looking at it's contents beyond that. I'd suggest replacing this with a boolean `passedOne` initialized to `false`, and later where you add to this list, instead set the boolean to `true`. Then when evaluating the overall result, just check for `singlePass && passedOne`.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (jimmyfagan) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:505]: How does `reverseFilter` work with `singlePass`? For condition filters it's not possible for them both to be true currently, but hypothetically if that were to happen, the logic here gets kind of confusing: `reverseFilter` reverses the result, but not if we find an exception, and not if `singlePass` is true and we find a passing condition.

**Review by jimmyfagan — COMMENTED** [review]
Left a handful of comments, some things you can probably update pretty easily, but overall I think `evaluateFilterCondition` is becoming a bit too complicated and doing too many different things depending on the parameters. I do like the general approach to just set a boolean to indicate whether to use and or or, so it might be best to just try to pair on this on Monday to see if we can figure out a clean way to break down that functionality.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt]: Yeah that makes sense. Updated!

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:407]: Been getting conflicting feedback about documentation, both for using `@param` and for using `[paramName]`. I think this needs to be tackled at some point via an official standard for the project

File: prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/RoutingTests.kt
Comment (TurkeyFried) [reply prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/RoutingTests.kt:1]: Yeah that's true! Added

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (jimmyfagan) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:407]: Sounds like an eng sync topic!

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:505]: I was also thinking of that one. I would think that the reverse of singlePass is `if any fail, it's false`

File: prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/RoutingTests.kt
Comment (jimmyfagan) [inline prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/RoutingTests.kt:292]:
```suggestion
            "Bundle.entry.resource.ofType(Provenance).count() > 50", // not true
```

**Review by jimmyfagan — COMMENTED** [review]
Thanks for the additions, it looks better! I'm still worried about some edge cases in `evaluateFilterCondition`, like when `reversed` is true or when the filter list is empty. Will try to huddle with you between meetings today, and hopefully we can work it out there!

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (TurkeyFried) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:505]: I'm gonna go with the latter of the above, and then we can go into more detail while discussing the overall complexity of the method

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (jimmyfagan) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt]: I think these functions might be a good example of why we shouldn't be too rigid with this, it's sometimes nice to explain params in-line, and it's sometimes better to explain generally what a function does, and then break down each param individually. We can optionally still chat about this in a future eng sync, but I'm good to resolve this conversation.

**Review by jimmyfagan — COMMENTED** [review]
We had a working session where we reworked some of this so that the evaluate condition function is not handling all combinations of "and/or" and "reversed/not-reversed", and instead we have two separate functions. I think that was a big improvement to readability and debugability, even if it is a little bit longer code overall given the separate functions. Holding off on approval til Luis checks out the code coverage of the split functions, and then will give my final review.

**Review by jimmyfagan — APPROVED** [review]
Might be good to add another test case for `reverseFilter=true` on the "or" path, but since that is currently an impossible case in the UP, I won't let that block my approval. If you add anything that warrants another look, let me know!

**Review by victor-chaparro — COMMENTED** [review]
Tested it locally and conditions were not getting filtered when I updated the conditionFilter for the development organization to use the new format.
The current filter is: 
```
      conditionFilter:
        # Accept all positive flu A and B results or negative flu A and B results for cities in LA and Orange counties
        - >-
          %resource.where(code.coding.code in ('85478-6' | '76080-1' | '92141-1' | '80383-3' | '82170-2' | '85477-8' | '76078-5' | '77026-3' | '77028-9' | '77027-1' | '92142-9' | '80382-5' | '82166-0' | '82167-8' | '82169-4' | '82168-6' | '92809-3' | '92808-5' | '95423-0')).value.coding.code.intersect('260373001' | '407479009' | '407480007').exists()
          or (%resource.where(code.coding.code in ('85478-6' | '76080-1' | '92141-1' | '80383-3' | '82170-2' | '85477-8' | '76078-5' | '77026-3' | '77028-9' | '77027-1' | '92142-9' | '80382-5' | '82166-0' | '82167-8' | '82169-4' | '82168-6' | '92809-3' | '92808-5')).value.coding.code.intersect('260415000').exists() and (%patient.address.city.intersect(%LACountyCities | %OrangeCountyCities).exists() or %performer.address.city.intersect(%LACountyCities | %OrangeCountyCities).exists()))
```
After updating it to use the new format the Covid conditions don't get filtered out:
 ```
     conditionFilter:
        # Accept all positive flu A and B results or negative flu A and B results for cities in LA and Orange counties
        - "%resource.where(code.coding.code in ('85478-6' | '76080-1' | '92141-1' | '80383-3' | '82170-2' | '85477-8' | '76078-5' | '77026-3' | '77028-9' | '77027-1' | '92142-9' | '80382-5' | '82166-0' | '82167-8' | '82169-4' | '82168-6' | '92809-3' | '92808-5' | '95423-0')).value.coding.code.intersect('260373001' | '407479009' | '407480007').exists()"
        - "(%resource.where(code.coding.code in ('85478-6' | '76080-1' | '92141-1' | '80383-3' | '82170-2' | '85477-8' | '76078-5' | '77026-3' | '77028-9' | '77027-1' | '92142-9' | '80382-5' | '82166-0' | '82167-8' | '82169-4' | '82168-6' | '92809-3' | '92808-5')).value.coding.code.intersect('260415000').exists() and (%patient.address.city.intersect(%LACountyCities | %OrangeCountyCities).exists() or %performer.address.city.intersect(%LACountyCities | %OrangeCountyCities).exists()))"
```

Attached is the postman collection I used to test this.

[CA condition test.txt](https://github.com/CDCgov/prime-reportstream/files/11502555/CA.condition.test.txt)

**Comment by jimmyfagan** [comment]
Checked in a new commit that should fix what Victor saw. There's still some questions I have about that section of code (I'll leave a comment and tag Jessica I think), and we may want to add some testing that would have caught that issue.

File: prime-router/src/main/kotlin/fhirengine/utils/FHIRBundleHelpers.kt
Comment (jimmyfagan) [inline prime-router/src/main/kotlin/fhirengine/utils/FHIRBundleHelpers.kt:1]: @JessicaWNava Why does `getObservationExtensions` return an empty list when `observationsToKeep.size == allObservations.size` and also when `observationsToKeep` is empty? They seem like very different scenarios that use the same behavior, which is a little bit confusing to me.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt
Comment (victor-chaparro) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:476]: Is this check necessary since the if before is checking if filter is null or empty?

Comment (jimmyfagan) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:476]: No it's not, that's leftover from an earlier iteration, we can remove it, good catch.

**Review by victor-chaparro — APPROVED** [review]
That was a quick fix! Re-tested it and I'm seeing the condition getting removed. I had written a ticket a while ago [here](https://app.zenhub.com/workspaces/platform-6182b02547c1130010f459db/issues/gh/cdcgov/prime-reportstream/8910) to add the integration test for removing conditions, but hasn't been groomed. I'll bring it up during our next grooming session.

**Comment by jimmyfagan** [comment]
@TurkeyFried Arnej said that he briefly looked through this, and without a formal approval, with me and Victor already approving it, we're okay to merge.

**Pull request merged**
